src/poetry/console/commands/add.py

```diff
@@ -215,6 +215,11 @@ def handle(self) -> int:
 
             constraint_name = _constraint["name"]
             assert isinstance(constraint_name, str)
+
+            for key in list(section.keys()):
```

Comment:
pedantry warning, none of this really matters

- if you're willing to assume that the `pyproject.toml` wasn't broken before we started, then 
```python
            for key in section:
                if canonicalize_name(key) == canonicalize_name(constraint_name):
                    del section[key]
                    break
```
should be sufficient, both avoiding the `list()` / `.keys()` ugliness, and avoiding doing more work than strictly necessary

- canonicalizing the name every time round the loop is ever so slightly inefficient, that could be pulled into a local variable

Reply:
Thanks a lot for your suggestions. I don't like the `list()` as well :smiley: